stdlib/tests/collections/smt.rs

```diff
@@ -0,0 +1,190 @@
+    let d = InsertedLeaf::new_seeded(&mut store, c.root, &mut seed, d_3, 48);
+    let e = InsertedLeaf::new_seeded(&mut store, d.root, &mut seed, e_3, 48);
+
+    // assert all nodes are returned with the MASM implementation
+    assert_leaves(store, e.root, [a, b, c, d, e]);
+}
+
+// TEST HELPERS
+// ================================================================================================
+
+/// Asserts key/value opens to root, provided the advice map and store
+fn assert_smt_get_opens_correctly(
+    key: Word,
+    value: Word,
+    root: Word,
+    store: MerkleStore,
+    advice_map: &[([u8; 32], Vec<Felt>)],
+) {
+    let source = r#"
+        use.std::collections::smt
+
+        begin
+            exec.smt::get
+        end
+    "#;
+    let initial_stack = [
+        root[0].as_int(),
+        root[1].as_int(),
+        root[2].as_int(),
+        root[3].as_int(),
+        key[0].as_int(),
+        key[1].as_int(),
+        key[2].as_int(),
+        key[3].as_int(),
+    ];
+    let expected_output = [
+        value[3].as_int(),
+        value[2].as_int(),
+        value[1].as_int(),
+        value[0].as_int(),
+        root[3].as_int(),
+        root[2].as_int(),
+        root[1].as_int(),
+        root[0].as_int(),
+    ];
+    let advice_stack = [];
+    build_test!(source, &initial_stack, &advice_stack, store, advice_map.iter().cloned())
+        .expect_stack(&expected_output);
+}
+
+/// Build the advice map and execute the test for all leaves, asserting they are returned by the
+/// MASM library
+fn assert_leaves<I>(store: MerkleStore, root: Word, leaves: I)
+where
+    I: IntoIterator<Item = InsertedLeaf>,
+{
+    let leaves: Vec<_> = leaves.into_iter().collect();
+    let advice_map: Vec<_> = leaves
+        .iter()
+        .map(|leaf| (leaf.node.into_bytes(), leaf.value.to_vec()))
+        .collect();
+
+    // tests all leaves individually
+    for leaf in leaves {
+        assert_smt_get_opens_correctly(leaf.key, leaf.value, root, store.clone(), &advice_map);
+    }
+}
+
+/// A representation of the post-insertion state of a leaf
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct InsertedLeaf {
+    /// Key
+    pub key: Word,
+    /// Remaining key for the depth
+    pub remaining_key: Word,
+    /// Generated value
+    pub value: Word,
+    /// Index in which the node was inserted
+    pub index: NodeIndex,
+    /// Computed node value
+    pub node: Word,
+    /// New root after insertion
+    pub root: Word,
+}
+
+impl InsertedLeaf {
+    /// Appends a new leaf with the provided path, returning the new root
+    pub fn new_seeded(
+        store: &mut MerkleStore,
+        root: Word,
+        seed: &mut u64,
+        path: u64,
+        depth: u8,
+    ) -> Self {
+        let mut leaf = Self::new_spurious(root, seed, path, depth);
+        leaf.root = store.set_node(leaf.root, leaf.index, leaf.node).unwrap().root;
+        leaf
+    }
```

Comment:
It is difficult for me to tell if this works correctly under the scenarios above. I think a more explicit approach would have been better (similar to what we had before but just encapsulated in a stand-alone function). But since it is already implemented, let's keep as is and create an issue to revisit these tests and improve in the future.

Reply:
These `new` functions are just boilerplate of generating a key/value pair, computing the remaining key, node value, etc.

Reply:
Right - but it is not clear to me how it resolve collisions key prefixes. Or does it assume that the caller will make sure to resolve such collisions?

Reply:
The prefix (namely `k[0, 1, 2]`) will be generated from the `seed`. It should be "virtually impossible" (or at least statistically unfeasible) to get a collision there.

The last limb of the key is defined by `path`, and it is expected from the caller as he should provide the correct bits to compose the test case.